sentry/sentry.go

```diff
@@ -0,0 +1,214 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package sentry
+
+import (
+	"os"
+	"time"
+
+	raven "github.com/getsentry/raven-go"
+	"github.com/pkg/errors"
+	"github.com/uber-go/zap"
+	"github.com/uber-go/zap/zwrap"
+)
+
+var (
+	_platform          = "go"
+	_traceContextLines = 3
+	_traceSkipFrames   = 3
+	_zapToRavenMap     = map[zap.Level]raven.Severity{
+		zap.InfoLevel:  raven.INFO,
+		zap.WarnLevel:  raven.WARNING,
+		zap.ErrorLevel: raven.ERROR,
+		zap.PanicLevel: raven.ERROR,
+		zap.FatalLevel: raven.FATAL,
+	}
+)
+
+// Logger automatically sends logs above a certain threshold to Sentry.
+type Logger struct {
+	zap.Meta
+	Capturer
+
+	// Minimum level threshold for sending a Sentry event
+	minLevel zap.Level
+
+	// Cotrols if stack trace should be automatically generated, and how many frames to skip
+	traceEnabled      bool
+	traceSkipFrames   int
+	traceContextLines int
+	traceAppPrefixes  []string
+
+	// Additional information that should be attached to each packet, i.e. customerUUID
+	extra zwrap.KeyValueMap
+}
+
+// Option pattern for logger creation.
+type Option func(l *Logger)
+
+// New Sentry logger.
+func New(dsn string, options ...Option) (*Logger, error) {
+	l := &Logger{
+		minLevel:          zap.ErrorLevel,
+		traceEnabled:      true,
+		traceSkipFrames:   _traceSkipFrames,
+		traceContextLines: _traceContextLines,
+		extra:             zwrap.KeyValueMap{},
+	}
+
+	for _, option := range options {
+		option(l)
+	}
+
+	// Default to live sentry packet capturer
+	if l.Capturer == nil {
+		client, err := raven.New(dsn)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create sentry client")
+		}
+
+		l.Capturer = &NonBlockingCapturer{Client: client}
+	}
+
+	return l, nil
+}
+
+// WithCapturer allows to specify which packet capturer should be used.
+func WithCapturer(c Capturer) Option {
+	return func(l *Logger) {
+		l.Capturer = c
+	}
+}
+
+// MinLevel provides a minimum level threshold, above which Sentry packtes will be triggered
+func MinLevel(level zap.Level) Option {
+	return func(l *Logger) {
+		l.minLevel = level
+	}
+}
+
+// TraceEnabled allows to change wether (the somewhat expensive, relatively)
+// stack trace generation takes place.
+func TraceEnabled(enabled bool) Option {
+	return func(l *Logger) {
+		l.traceEnabled = enabled
+	}
+}
+
+// TraceCfg allows to change the number of skipped frames, number of context lines and
+// list of go prefixes that are considered "in-app", i.e. "github.com/uber-go/zap".
+func TraceCfg(skip, context int, prefixes []string) Option {
+	return func(l *Logger) {
+		l.traceSkipFrames = skip
+		l.traceContextLines = context
+		l.traceAppPrefixes = prefixes
+	}
+}
+
+// Extra stores additional information for each Sentry event.
+func Extra(extra map[string]interface{}) Option {
+	return func(l *Logger) {
+		l.extra = extra
+	}
+}
```

Comment:
For congruence with `zap.Fields`, consider calling this `Fields` and taking a variadic number of `zap.Fields` as arguments. We'll likely want to pass the same information in both places, so representing it the same way will be convenient.